### ISSUE TITLE
AppSourceCop - How to deal with previous app versions

### ISSUE THREAD

**Issue by fvet** [HodorNV/ALOps#234]
**Is your feature request related to a problem? Please describe.**
Just taking our first steps towards an AppSource proof pipeline.

Seems like the new Run-ALPipeline provide support for AppSourceCop validation using the **previous** app versions.
https://freddysblog.com/2020/10/08/run-alpipeline/
![image](https://user-images.githubusercontent.com/25268332/97183879-9232b380-179e-11eb-81cf-ad6ae087810d.png)

Also, for development, we'll have to ensure the previous app version is located somewhere in the packachecache.
https://docs.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/analyzers/appsourcecop
![image](https://user-images.githubusercontent.com/25268332/97183949-a70f4700-179e-11eb-8ea1-43bd4446ae52.png)

How can ALOps simplify this process (knowing we might have x apps and will release every x months)? 
Any sample pipelines available on how to easily reference to a 'Previous App' version? Do you take the apps from a previous pipeline (and harcode the reference), do you include the app in your repo, ...? Any native support will be provided in some of the ALOps tasks?

**Comment by waldo1001** [comment]
Enabling AppSourceCop has been possible for a long time: use the al_analyzer param.

You can already introduce a breaking change test by downloading the previous artifacts (same build), and publish your new version over the old one:
https://www.youtube.com/watch?v=CCJG9cnjmLs&t=3392s

We have the breaking change test on the roadmap for CompilerV2, but this would still mean you have to provide the previous version by downloading the artifacts.  You can expect an extra setting to provide previous version path of the app.. .

**Comment by MortenRa** [comment]
> You can expect an extra setting to provide previous version path of the app.. .

I´m in doubt how to configure this when using templates where the DownloadBuildArtifacts step is configured, do you have some examples, and will the extra setting be somethng in a new build

**Comment by fvet** [comment]
Thanks Eric, was aware of al_analyzer and AppSourceCop, but still have some questions on best practices on how to reference 'previous' app versions.

Look at the vid, you use the latest succesfull build artifacts from a build pipeline. This would detect breaking changes on each individual PR:
1. Create object > PR + build > create version 1 of app
2. Add field to object > PR + build > download version 1 of app > create version 2 of app
3. Extend field from previous step > PR + build > download version 2 of app > this would be considered a breaking change, leaving us with no option than obsoleting a field (that never even got into production). (since the code reviewer might have overlooked it) Is this correct or do you apply any workaround?

In the above case, the breaking change seems not critical to us as long as it has not been 'released' to any environment. (we currently don't have release pipelines that are PR triggered).

The above also tests seems good to detect breaking changes between version and version - 1 on PR level. Do you apply the same rules for Appsource? For Appsource we've submitted a specific version (1) of the app and will only deploy a newer version (10) within a longer period of time. Do you test against the originally submitted version (1) or do you consider the tests on all individual PRs / accumulating deltas to be safe as well?

Are the breaking changes only part of the CI process / PR branch validation by installing the previous version only then, or do you include the previous (or AppSource) version in the appsourcecop.json file so developers will be alerted during development already? (if so, the artifacts need to be downloaded to the .alpackages folder indivually)

**Comment by fvet** [comment]
@AdminHodor @waldo1001 
The setup we're currently testing is the folllowing:

**Prequisites**
- Version 1.0.0.0 of app, submitted on AppSource

**VS Code**
- Add version1.0.0.0.app in .alpackages folder (onetime upload, since folder is normally gitignored)
- Add version : "1.0.0.0" in the AppSourceCop.json > we want to detect breaking changes against the app that is currenty on Appsource
- Enable "AppSourceCop" codeanalyzer
- Update app.json to version "2.0.0.0"
- Developer is aware of any breaking changes compared to version "1.0.0.0" upon compiling / packaging > OK

**CI / ALOPS**
- ALOpsCompile@1 : Fails, since the PackageCache folder does not contain version1.app (since there's no dependencies in the app.json, so even publishing version1.0.0.0 before does not makes any difference) 

```
C:\Run\Microsoft.al\al-ALLanguage\extension\bin\alc.exe /packagecachepath:"c:\Run\DevOps\PackageCache" 
...
##[error]c:\Run\DevOps\app\AppSourceCop.json(1,1): error AS0003: The version '1.0.0.0' of the extension with publisher 'Navitrans' and name 'Core' could not be found in the package cache folder. If you do not want to validate breaking changes, remove the property 'version' from the AppSourceCop.json.
```

In another attempt, we've tried to move the .alpackages content to the PackageCache folder in the working directory first, before it's copied to the container (c:\Run\DevOps) ...

![image](https://user-images.githubusercontent.com/25268332/97971008-ae4bdb80-1dc2-11eb-9d90-b78dc794e415.png)

but without any luck. Apparently the ALOpsCompile step has cleaned the c:\Run\DevOps\PackageCache?

- ALOpsCompile@2 : Seems to work somewhat better - as it scans all .app files from the workingdir and uses them for compiling - but we're blocked on various other pending issues in case of multiple apps / repo's combined (incl. signing), forcing us to stay with ALOpsCompile@1

In both ways we're stuck if a version is referenced in the AppSourceCop.json, since the compiler seems unaware of the previous app file. Any change in having this prioritized / setup a Teams meeting for more details?

**Comment by fvet** [comment]
@waldo1001 Is the appsourceversion check you've implemented internally part of the 'VS Code' experience (e.g. developers are alerted before creating a PR because the version is part of the appsourcecop.json file). Or does ALOps includes the previous version of the appsourcecop.json file only as part of the CI pipeline?

Not sure how to interpret this... Developers is coding blind without relying on the appsourcecop and will only get warnings when the PR has been created.

![image](https://user-images.githubusercontent.com/25268332/100370165-2abe8c80-3006-11eb-8fcd-79f8badb1a0e.png)

**Comment by fvet** [comment]
Abandoning the Appsourcecop previous version check in VS Code currently seems to be the best option to move forward, but then we need ALOps to be able to implement the previous version checks.

Can we imagine having a setting pointing to previous apps (could be part of a repo .PreviousApps subfolder, universal package, pipeline artifacts, external zip, ...) ...

![image](https://user-images.githubusercontent.com/25268332/100371238-c56b9b00-3007-11eb-933f-5530891e7811.png)

... and have ALOps updating the appsourcecop as part of the build process (taking over the VS Code behavior untill MS comes up with a decent fix)?

![image](https://user-images.githubusercontent.com/25268332/100371264-cf8d9980-3007-11eb-9d29-2498806fc0a4.png)

https://github.com/microsoft/navcontainerhelper/blob/0dc4d65b9b290f873a7f0c8a0714ef458c98281c/AppHandling/Run-AlPipeline.ps1#L754

**Comment by fvet** [comment]
Keeping the previous app files as part of your repo / workingdirectory seems not a a good idea when doing a (batch) compile (e.g. to  prepate the latest g.xliffs and sync the target languages for all apps)

Steps:
- Checkout all repo's
- Compile all apps

```
- task: ALOpsAppCompiler@2
            displayName: Compile applications
            inputs:
              artifactversion: ${{ parameters.artifactVersion }}
              artifacttype: OnPrem
              artifactcountry: ${{ parameters.country }}
              appversiontemplate: '?.?.?.*'
              alcodeanalyzer: ''
              failonwarnings: false
              publishartifact: false
              alsourcepath: $(System.DefaultWorkingDirectory)
```

- failure on one repo containing a previous app in .alpackages (sourcecontroled since currently no other samples from ALOps on how to include the file for AppSourcecop tests)

```
 -select: Latest -version: 17.1 -type: OnPrem -country: BE
*** Platform: c:\bcartifacts.cache\onprem\17.1.18256.18792\platform
*** Localisation: c:\bcartifacts.cache\onprem\17.1.18256.18792\be
*** BC Management DLL: C:\bcartifacts.cache\onprem\17.1.18256.18792\platform\LegacyDlls\program files\Microsoft Dynamics NAV\170\RoleTailored Client\Microsoft.Dynamics.Nav.Management.dll
*** BC App Management DLL: C:\bcartifacts.cache\onprem\17.1.18256.18792\platform\LegacyDlls\program files\Microsoft Dynamics NAV\170\RoleTailored Client\Microsoft.Dynamics.Nav.Apps.Management.dll
*** Get [*.app] and [app.json] files recursive from [C:\DockerInstallation\Agent\_work\121\s]
*** Found [51] files.
*** Extract App detail from [C:\DockerInstallation\Agent\_work\121\s\Navitrans.Ameta\app\app.json]
*** Extract App detail from [C:\DockerInstallation\Agent\_work\121\s\Navitrans.bAlert\app\app.json]
*** Extract App detail from [C:\DockerInstallation\Agent\_work\121\s\Navitrans.Base\app\app.json]
*** Extract App detail from [C:\DockerInstallation\Agent\_work\121\s\Navitrans.Base\test\app.json]
*** Extract App detail from [C:\DockerInstallation\Agent\_work\121\s\Navitrans.BC\app\app.json]
*** Extract App detail from [C:\DockerInstallation\Agent\_work\121\s\Navitrans.BE\app\app.json]
*** Extract App detail from [C:\DockerInstallation\Agent\_work\121\s\Navitrans.BE\test\app.json]
*** Extract App detail from [C:\DockerInstallation\Agent\_work\121\s\Navitrans.BulkSMS\app\app.json]
*** Extract App detail from [C:\DockerInstallation\Agent\_work\121\s\Navitrans.Carcube\app\app.json]
*** Extract App detail from [C:\DockerInstallation\Agent\_work\121\s\Navitrans.Cleaning\app\app.json]
*** Extract App detail from [C:\DockerInstallation\Agent\_work\121\s\Navitrans.ContainerDispatch\app\app.json]
*** Extract App detail from [C:\DockerInstallation\Agent\_work\121\s\Navitrans.Core\.alpackages\Navitrans_Navitrans 365 Core_17.0.0.6562_OnPrem_17.1_w1.app]
##[error]The specified file is not an extension file.

```

**Comment by waldo1001** [comment]
In CompilerV2 this already works (more info here: https://www.waldo.be/2020/12/04/run-your-own-appsource-validation/)

We'll add it in CompilerV1 as well.

**Comment by fvet** [comment]
Opened a new issue on how to achieve batch validation (https://github.com/HodorNV/ALOps/issues/279)

How do you manage the updates of the app files in the .previous folder and the version in the appsourcecop.json when releasing ? (e.g tens of apps, released every x months, would require some automation to manage the previous apps)

I was thinking of adding a step in our release pipeline when the apps are officially 'released'/published as universal packages (or to another storage*), to also replace/commit the new apps in the .previous folder for each app / repo and update the version in the appsourcecop.json (considering the new released version of the apps will be submitted to AppSource).

It would - again - result in a bunch of PR's, where maybe it would be beneficial to be able to:
- Host our previous version of all our app files on a separate storage (cfr *)
- Download these files as part of the AppSourceValidation pipeline (and move to .previous, once per app)
- Update the appsourcecop.json with the version number from the .previous app (similar to bccontainerhelper)

This still leaves a gap for the VSCode part, where the above step would need a PS equivalent for the developer to update his workspace. So I understand keeping the apps part of the repo, but am looking how to easily include updating these apps upon releasing new versions.

**Comment by fvet** [comment]
We have a working setup allowing us to include the previous released version as part of the repo and update the AppSourceCop.json version (conditionally updated as part of the release pipeline). The ALOps compiler does pickup these apps to test on breaking changes and the developer has the apps as well when pulling the latest code. 

So for us, the scenario works.

Issue closed ;)

**Comment by pri-kise** [comment]
@fvet would you mind to share how you update your Repo conditionally with the artifact in your release pipeline?

**Comment by fvet** [comment]
@pri-kise At my current employer, we still manage this manually. In the past, I remember we had added a step in the release pipeline to take the compiled app files and copy them to a subfolder in the repo and update the AppSourceCop.json file.

Here's some inspiration (not all elements needed are included).

Still a pity that ALOps does not contain anything out of the box to simplify this process...

```
stages:
  - stage: update_previous_apps
    displayName: Update Previous Apps
    jobs:
      - job: update_previous_apps
        displayName: Update Previous Apps
        steps:
          - checkout: git://Scripts
            persistCredentials: true          
          - ${{ if eq(parameters.bundle, 'All') }}:
            - template: CI Pipelines\clone_repositories_full.yml@templates          

          - task: PowerShell@2
            displayName: Change to target branch
            inputs:
              filePath: $(System.DefaultWorkingDirectory)\Scripts\Development\Git\Change branches.ps1
              arguments: -WorkingDirectory '$(System.DefaultWorkingDirectory)' -Branch '${{ parameters.targetBranch }}'
          - task: PowerShell@2
            displayName: Change to working branch
            inputs:
              filePath: $(System.DefaultWorkingDirectory)\Scripts\Development\Git\Change branches.ps1
              arguments: -WorkingDirectory '$(System.DefaultWorkingDirectory)' -Branch 'prepUpdatePreviousApps'

          - task: PowerShell@2
            displayName: Clear PreviousApps Folder
            inputs:
              targetType: 'inline'
              script: |
                if (Test-Path -Path '$(System.DefaultWorkingDirectory)\PreviousApps') {
                    Remove-item -Path '$(System.DefaultWorkingDirectory)\PreviousApps' -Recurse -Force
                }

          - template: CI\download_artifacts\previous\full.yml@templates
            parameters:
              previous_version: ${{ parameters.previousApps }}                            
              remove_testapps: false
              overrule_country: true

          - task: PowerShell@2
            displayName: Update Previous Apps
            inputs:
              filePath: $(System.DefaultWorkingDirectory)\Scripts\Development\Devops\UpdatePreviousApps.ps1
              arguments: -WorkingDirectory '$(System.DefaultWorkingDirectory)' -Message '${{ parameters.message }}' -NextMajorMinor '${{ parameters.nextMajorMinor }}'
              
          - task: PowerShell@2
            displayName: Create pull requests
            inputs:
              filePath: $(System.DefaultWorkingDirectory)\Scripts\Development\Git\Create pull requests.ps1
              arguments: -WorkingDirectory '$(System.DefaultWorkingDirectory)' -Message '${{ parameters.message }}' -SourceBranch 'prepUpdatePreviousApps' -TargetBranch '${{ parameters.targetBranch }}' -RepositoryLogPath '$(System.DefaultWorkingDirectory)\.temp\repositories.log'
            env:
              AUTHORIZATION_TYPE: Bearer
              AUTHORIZATION_TOKEN: $(System.AccessToken)
```


```
[CmdletBinding()]
param (
    [Parameter(Mandatory = $true)]
    [string]$WorkingDirectory,
    [Parameter(Mandatory = $true)]
    [string]$Message,
    [Parameter(Mandatory = $false)]
    [string]$NextMajorMinor
)

function Get-ScriptsRoot() {
    $Root = $PSScriptRoot
    $Name = Split-Path -Path $Root -Leaf

    while ($Name -ne 'Scripts') {
        $Root = Split-Path -Path $Root -Parent
        $Name = Split-Path -Path $Root -Leaf
    }

    return $Root
}

function Get-Repositories() {
    param (
        $ParentFolder
    )
    Get-ChildItem -Path $ParentFolder -Directory -Depth 1 -Filter '.git' -Force `
        | Where-Object { $_.Parent.Name -ne 'Scripts' } `
        | ForEach-Object { $_.Parent }    
}

function Process-Repository() {
    param (
        $Repository,
        [Parameter(Mandatory = $false)]
        [string]$NextMajorMinor
    )
    Write-Host "Repository => $($Repository.Name)"

    Set-Location -Path $Repository.FullName

    $AppSourceApps = (Get-ChildItem -Path '.' -Filter 'AppSourceCop.json' -Depth 2)
    if (!($AppSourceApps)) {
        Write-Host "`tNo AppSourceCop.json found"
    } else {
        $ChangeApplied = $false
        foreach ($AppSourceApp in $AppSourceApps) {
           Write-Host "`tAppSourcecop Found in" $AppSourceApp.Directory.Name
           $ChangeApplied = $ChangeApplied -or (UpdatePreviousApp -AppFolder $AppSourceApp.Directory.Name -NextMajorMinor $NextMajorMinor)
        }
        if ($ChangeApplied) {
            Write-Host "`tCommit & push changes to prepUpdatePreviousApps"
            Add-Content -Path $RepositoryLogPath -Value $Repository.Name
            git commit -q -m $Message *> $null
            git push -q -u origin prepUpdatePreviousApps *> $null            
        }
    }
    Write-Host
}

function UpdatePreviousApp() {
    param (
        $AppFolder,
        [Parameter(Mandatory = $false)]
        [string]$NextMajorMinor
    )
    $AppPath = '.\' + $AppFolder
    if (!(Test-Path "$AppPath\app.json")) {
        Write-Host "No $AppType found"
        return $false
    }
    $AppInfo = (Get-Content -Raw -Path "$AppPath\app.json" | ConvertFrom-Json)
    $AppName = $AppInfo.name
    $AppFilter = '*_' + $AppName + '_*.app'
    $App = (Get-ChildItem -Path "$WorkingDirectory\PreviousApps" -Filter "$AppFilter" -Recurse)
    if (!$App) {
        Write-Host "`t`tNo Previous Release Found for $AppName"
        return $false
    }
    if (Test-Path "$AppPath\.alpackages\*") {
        Get-ChildItem -Path "$AppPath\.alpackages" | ForEach-Object {
            $FileToRemove = (Join-Path -Path "$AppPath\.alpackages" -ChildPath $_.Name)
            Write-Host "`t`tRemoving $FileToRemove"
            git rm -f $FileToRemove *> $null
        }
    }
    $DestinationFile = (Join-Path -Path "$AppPath\.alpackages" -ChildPath $App.Name)
    Write-Host "`t`tAdding $DestinationFile"
    New-Item $DestinationFile -Force | Out-Null
    Copy-Item -Path $App.FullName -Destination $DestinationFile -Recurse -Force | Out-Null
    git add -f $DestinationFile *> $null

    $AppSourceVersion = ($App.Name -split '_')[2]
    $AppSourceCopFile = Join-Path -Path $AppPath -ChildPath 'AppSourceCop.json'
    $AppSourceCopContent = (LoadFrom-Json -FilePath $AppSourceCopFile)
    
    Update-Property -InputObject $AppSourceCopContent -Property 'version' -Value $AppSourceVersion
    if ($null -ne $NextMajorMinor) {
        Update-Property -InputObject $AppSourceCopContent -Property 'obsoleteTagVersion' -Value $NextMajorMinor
    }

    SaveTo-Json -InputObject $AppSourceCopContent -FilePath $AppSourceCopFile
    git add $AppSourceCopFile *> $null
    return $true
}

function LoadFrom-Json() {
    param (
        [Parameter(Mandatory = $true)]
        [string]$FilePath
    )

    $Settings = [Newtonsoft.Json.Linq.JsonLoadSettings]::new()
    $Settings.CommentHandling = 'Load'
    
    $Content = Get-Content -Path $FilePath -Raw -Encoding 'UTF8'
    $OutputObject = [Newtonsoft.Json.Linq.JObject]::Parse($Content, $Settings)

    return ,$OutputObject
}

function Update-Property() {
    param (
        [Parameter(Mandatory = $true)]
        [object]$InputObject,
        [Parameter(Mandatory = $true)]
        [string]$Property,
        [Parameter(Mandatory = $false)]
        [object]$Value
    )

    $Updated = $true
    $Token = $InputObject[$Property]

    if ($null -ne $Token) {
        $Updated = $Token.Value -ne $Value
        $Token.Value = $Value
    } else {
        $InputObject[$Property] = [Newtonsoft.Json.Linq.JToken]::FromObject($Value)
    }

    return $Updated
}

function SaveTo-Json() {
    param (
        [Parameter(Mandatory = $true)]
        $InputObject,
        [Parameter(Mandatory = $true)]
        [string]$FilePath
    )

    $FileStream = $null
    $StreamWriter = $null
    $JsonWriter = $null

    try {
        $FileStream = [System.IO.File]::Open($FilePath, 'Create')
        $StreamWriter = New-Object System.IO.StreamWriter $FileStream
        $JsonWriter = New-Object Newtonsoft.Json.JsonTextWriter $StreamWriter
        $JsonWriter.Formatting = 'Indented'
        $JsonWriter.IndentChar = ' '
        $JsonWriter.Indentation = 2

        if ($InputObject.GetType().IsSubclassOf([Newtonsoft.Json.Linq.JToken])) {
            $InputObject.WriteTo($JsonWriter)
        } else {
            $JsonWriter.WriteStartObject()
            $JsonWriter.WriteValue($InputObject)
            $JsonWriter.WriteEndObject()
        }
    }
    finally {
        if ($null -ne $JsonWriter) {
            $JsonWriter.Dispose()
        }

        if ($null -ne $StreamWriter) {
            $StreamWriter.Dispose()
        }

        if ($null -ne $FileStream) {
            $FileStream.Dispose()
        }
    }
}

if ([string]::IsNullOrWhiteSpace($NextMajorMinor)) {
    $NextMajorMinor = $null
}

$ScriptsRoot = Get-ScriptsRoot
[Reflection.Assembly]::LoadFile("$ScriptsRoot\Libs\Newtonsoft.Json@12.0.3\net45\Newtonsoft.Json.dll")

$TempDirectory = [System.IO.Path]::Combine($WorkingDirectory, '.temp')
New-Item -Path $TempDirectory -ItemType 'directory' -ErrorAction SilentlyContinue | Out-Null
$RepositoryLogPath = [System.IO.Path]::Combine($TempDirectory, 'repositories.log')
New-Item -Path $RepositoryLogPath -ItemType 'file' -Force -ErrorAction SilentlyContinue | Out-Null

$Repositories = (Get-Repositories -ParentFolder $WorkingDirectory)
foreach ($Repository in $Repositories) {
    Process-Repository -Repository $Repository -NextMajorMinor $NextMajorMinor
}
```

**Comment by waldo1001** [comment]
I'm all ears to look into this to simplify .. but I have feeling it's going to be something link opening Pandora's box: changing the repo from a pipeline is something I wouldn't want to have too much responsibility for, and is usually highly dependent of how people manage their projects/repos/... .

So I have a feeling it's a bunch of scripting that cannot be generalized, but only has a few challenges:
- change files in a repo
- commit to repo
But .. with lots of parameters:
- one app?  batch?
- appname vs repo name
- what are test-apps?
- include/exclude
- one app one repo, or one repo multiple apps?

in our case, it's "just" a (very) custom script we only run when we release to AppSource...

We'll look into it, but no promises.
